npm/design-system/src/components/FileExplorer/FileExplorer.tsx

```diff
@@ -0,0 +1,177 @@
+  isOpen: boolean
+  onClick: () => void
+}
+
+export interface FileComponentProps {
+  item: FileNode
+  depth: number
+  onClick: (file: FileNode) => void
+}
+
+export interface FileExplorerProps extends React.HTMLAttributes<HTMLDivElement> {
+  files: TreeNode[]
+  fileComponent: React.FC<FileComponentProps>
+  folderComponent: React.FC<FolderComponentProps>
+  selectedFile?: string
+  onFileClick: (file: FileNode) => void
+
+  // Styles. They should be a *.module.scss.
+  // TODO: Can we type these? Do we want to couple to CSS modules?
+  cssModule?: {
+    nav: any
+    ul: any
+    li: any
+    a: any
+    isSelected: any
+  }
+}
+
+export interface FileTreeProps extends FileExplorerProps {
+  depth: number
+  openFolders: Record<string, boolean>
+  style?: React.CSSProperties
+  setSelectedFile: (absolute: string) => void
+}
+
+export const FileExplorer: React.FC<FileExplorerProps> = (props) => {
+  /**
+   * Whether a folder is open or not is a **UI** concern.
+   * From a file system point of view, there is no such concept as "open" or "closed",
+   * only from a user's point of view.
+   * For this reason we save the open state as part of the UI component. The easiest
+   * way to do this is a key/value pair, mapping the absolute path of a directory to a boolean
+   *
+   * {
+   *   'foo': true,
+   *   'foo/bar': true
+   *   'foo/bar/qux': false
+   * }
+   *
+   * Every directory is set to open by default. When you add a new directory
+   * or file via your file system (eg mkdir foo/bar && touch foo/bar/hello.js) it will be added
+   * without losing the current state of open/closed directories.
+   */
+  const [openFolders, setOpenFolders] = React.useState<Record<string, boolean>>({})
+
+  React.useEffect(() => {
+    function walk (nodes: TreeNode[]) {
+      for (const node of nodes) {
+        if (node.type === 'folder') {
+          if (!(node.absolute in openFolders)) {
+            setOpenFolders({ ...openFolders, [node.absolute]: true })
+          }
+
+          walk(node.files)
+        }
+      }
+    }
+
+    walk(props.files)
+  }, [props.files, openFolders])
+
+  const setSelectedFile = (absolute: string) => {
+    setOpenFolders({ ...openFolders, [absolute]: !openFolders[absolute] })
+  }
+
+  return (
+    <nav className={cs(props.className, props.cssModule.nav)}>
+      <FileTree
+        {...props}
+        setSelectedFile={setSelectedFile}
+        openFolders={openFolders}
+        depth={0}
+      />
+    </nav>
+  )
+}
+
+export const FileTree: React.FC<FileTreeProps> = (props) => {
+  // Negative margins let the <a> tag take full width (a11y)
+  // while the <li> tag with text content can be positioned relatively
+  // This gives us HTML + cssModule-only highlight and click handling
+  const inlineStyles = {
+    a: {
+      marginLeft: `calc(-20px * ${props.depth})`,
+      width: `calc(100% + (20px * ${props.depth}))`,
+    },
+    li: {
+      marginLeft: `calc(20px * ${props.depth})`,
+    },
+  }
```

Comment:
Unless I'm mistaken, both of these fields are only consumed once. Why are they not defined inline?

Also, please actually calculate the positions in advance if possible. No need to use `calc` in situations where we know the sizes.

Reply:
Done... we still have one `calc`, but it'll do for now until we replace it with something more optimize.